### PR TITLE
Results schema

### PR DESCRIPTION
**Pull request by litzenberger**
Added schema and validation
Added file out
Added unit tests

### CODE REVIEW

File: tests/local-add.js [hunk 1]
```diff
@@ -4,36 +4,36 @@
...
-        file: file,
-        date: new Date().toISOString(),
-        s: end[0],
-        ms: end[1] / 1000000,
-        cpu: os.cpus(),
-        loadAvg: os.loadavg(),
-        memory: os.totalmem() - os.freemem()
-      }
-    )
+    let model = generateModel()
```
Comment (mcollina) [inline tests/local-add.js]: should be `const`

File: tests/schema/results.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env node
```
Comment (mcollina) [inline tests/schema/results.js]: I don't think this is executable.

File: tests/schema/results.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  .prop('duration')
+  .ref('#definitions/duration')
+  .definition(
+    'meta',
+    FluentSchema()
+      .prop('project', FluentSchema()
+        .default('js-ipfs'))
+      .prop('commit')
+      .prop('version')
+    // .required()
```
Comment (mcollina) [inline tests/schema/results.js]: why this is commented?

File: tests/schema/results.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  },
+  'cpu': 'cpu',
+  'loadAvg': 'load average',
+  'memory': 'memory'
+}
+
+function resultModel () {
+  let user = {}
+  const ajv = new Ajv({ useDefaults: true })
+  const validate = ajv.compile(schema.valueOf())
```
Comment (mcollina) [inline tests/schema/results.js]: We should be storing this `validate` function into the global scope.

File: tests/schema/results.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  validate(user)
+  return user
+}
+
+function generateModel () {
+  return model
+}
```
Comment (mcollina) [inline tests/schema/results.js]: Why is this function needed at all?

File: tests/unit/output.test.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+test('validate - results ', t => {
+  t.plan(2)
+  try {
+    const e = validate(resultModel())
+    t.equal(e, true)
+    t.pass()
+  } catch (e) {
+    console.log(e)
+    t.fail()
+  }
```
Comment (mcollina) [inline tests/unit/output.test.js]: Can you just remove the try catches? They are not needed.

File: tests/lib/output.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const folder = 'out'
+async function write (data) {
+  const name = createFilename(data)
+
+  if (validate(data)) {
+    try {
+      fs.writeFileSync(`${name}.json`, JSON.stringify(data))
+      console.log(data)
+    } catch (e) {
+      throw new Error(e)
```
[the remark this answers is not in the document]
Comment (litzenberger) [reply tests/lib/output.js]: removing

File: tests/local-add.js (hunk 1 above)
Comment (litzenberger) [reply tests/local-add.js]: yes.  updating.

File: tests/schema/results.js (hunk 5 above)
Comment (litzenberger) [reply tests/schema/results.js]: its a placeholder was trying to generate the model using AJv but not all the defaults were being populated.  Will remove that until I can get Ajv to generate it.

File: tests/unit/output.test.js (hunk 6 above)
Comment (litzenberger) [reply tests/unit/output.test.js]: remove

File: tests/local-add.js [hunk 8]
```diff
@@ -4,36 +4,36 @@
...
+    model.file = file
+    model.date = new Date().toISOString()
+    model.description = 'Add file to local repo'
+    model.testClass = testClass
+    model.duration.s = end[0]
+    model.duration.ms = end[1] / 1000000
+    model.cpu = os.cpus()
+    model.loadAvg = os.loadavg()
+    model.memory = os.totalmem() - os.freemem()
+    return model
```
[the remark this answers is not in the document]
Comment (litzenberger) [reply tests/local-add.js]: yes, that is already populated

[the remark this answers is not in the document]
Comment (litzenberger) [reply tests/local-add.js]: https://github.com/ipfs/benchmarks/blob/3b0e9d1b16a97a2ccb514c58d8b16542670ae265/tests/schema/results.js#L67

File: tests/lib/output.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  }
+}
+
+async function createFilename (data) {
+  try {
+    await fsExists(folder)
+    return `${folder}/${data.name}-${new Date().toISOString()}`
+  }
+  catch (err) {
+    console.log(err)
```
Comment (mcollina) [inline tests/lib/output.js]: this `console.log`  is spurious.

File: tests/lib/output.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+const schema = require('../schema/results.js')
+
+const folder = 'out'
+async function write (data) {
+  const name = await createFilename(data)
+
+  if (validate(data)) {
+    await fsWriteFile(`${name}.json`, JSON.stringify(data))
+    console.log(data)
+  }
```
Comment (mcollina) [inline tests/lib/output.js]: if this does not validate?

File: tests/schema/results.js [hunk 11]
```diff
@@ -79,10 +78,10 @@ const resultsDTO = {
   'memory': 'memory'
 }
 
+const validate = ajv.compile(schema.valueOf())
+
 function resultModel () {
   let user = {}
```
[the remark this answers is not in the document]
Comment (litzenberger) [reply tests/schema/results.js]: its from a test I was doing.  removing

File: tests/lib/output.js (hunk 10 above)
Comment (litzenberger) [reply tests/lib/output.js]: yep, adding a test for that case.

File: tests/local-extract.js [hunk 12]
```diff
@@ -1,30 +1,32 @@
...
+    model.subtest = subtest
+    model.file = file
+    model.date = new Date().toISOString()
+    model.description = 'Get file to local repo'
+    model.testClass = testClass
+    model.duration.s = end[0]
+    model.duration.ms = end[1] / 1000000
+    model.cpu = os.cpus()
+    model.loadAvg = os.loadavg()
+    model.memory = os.totalmem() - os.freemem()
```
Comment (mcollina) [inline tests/local-extract.js]: I would very much prefer a model where:

```js
return build({
  // props
})
```

File: tests/lib/output.js (hunk 10 above)
Comment (litzenberger) [reply tests/lib/output.js]: invalid files are written out to error directory.

**Pull request merged**